File: csharp/circular-buffer/CircularBuffer.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class CircularBuffer<T>(int capacity)
{
    private int Capacity { get; } = capacity;
    private List<T> Items { get; } = new(capacity);

    public T Read()
    {
        if (Items.Count == 0)
        {
            throw new InvalidOperationException();
        }

        T item = Items.First();
        Items.RemoveAt(0);

        return item;
    }

    public void Write(T value)
    {
        if (Items.Count == Capacity)
        {
            throw new InvalidOperationException();
        }

        Items.Add(value);
    }

    public void Overwrite(T value)
    {
        if (Items.Count == Capacity)
        {
            Items.RemoveAt(0);
        }

        Write(value);
    }

    public void Clear()
    {
        if (Items.Count == 0)
        {
            return;
        }

        Items.Clear();
    }
}

File: kotlin/circular-buffer/src/main/kotlin/CircularBuffer.kt
import kotlin.collections.ArrayDeque

class EmptyBufferException: Exception("The buffer is empty.")

class BufferFullException: Exception("The buffer is full.")

class CircularBuffer<T>(size: Int) {
    val initialCapacity: Int = size
    val array: ArrayDeque<T> = ArrayDeque(initialCapacity)

    fun read() : T {
        if (array.isEmpty()) {
            throw EmptyBufferException()
        }

        return array.removeFirst()
    }

    fun write(value: T) {
        if (array.size == initialCapacity) {
            throw BufferFullException()
        }

        array.addLast(value)
    }

    fun overwrite(value: T) {
        if (array.size == initialCapacity) {
            array.removeFirstOrNull()
        }

        write(value)
    }

    fun clear() {
        array.clear()
    }
}
